Add permalinks for Web Literacy Map expander sections

toolness opened the pull request:
This is a potential fix for #975. It's a bit weird because our Expander already actually behaves somewhat inconsistently, and making it accessible is not particularly straightforward.

Right now this adds a [section sign (§)](http://en.wikipedia.org/wiki/Section_sign) at the top-right of each expander section. Clicking on it will set your browser's current URL to a permalink for the section. Here's an example permalink to the Accessibility section of the Web Literacy Map, hosted on my personal deployment branch thingy:

http://mozteach.toolness.org/teach-like-mozilla/web-literacy/#accessibility

Visiting that *should* automatically expand the Accessibility section and focus keyboard navigation on it.

(You can also see what that link looks like [with JS disabled](http://mozteach.toolness.org/teach-like-mozilla/web-literacy/?safemode=on#accessibility).)

Right now this PR throws out a lot of the focus/blur stuff in the existing code because (A) it was buggy/not behaving consistently and (B) trying to make it un-buggy resulted in lots of weird annoying edge cases, such as always collapsing the current section when switching to a different application.

ottonomy commented:
Functionally, :+1: . The only thing I noticed it didn't do is if I entered a new anchor in the URL bar, it didn't expand the new section. That's an edge case though, and I don't know if it's really worth adding a handler to catch it.

Visually, one pattern that you might be interested in is like GitHub does for wiki pages, where the permalink icon is only visible when hovering/focusing the target. You can see it if you hover over the headings on [this page](https://github.com/mozilla/openbadges-specification/blob/master/Assertion/latest.md) for instance.

Thanks for getting on this so quickly! I'm excited to badge the web lit map. :)

toolness commented:
Thanks for the feedback @ottonomy!

> The only thing I noticed it didn't do is if I entered a new anchor in the URL bar, it didn't expand the new section. That's an edge case though, and I don't know if it's really worth adding a handler to catch it.

Yeah, I actually originally had event listeners attached to `hashchange` during component mount/unmount, but I felt a bit weird about attaching so many listeners... I also want to add exhaustive tests for all this, and doing so for `hashchange` will take time--when I realized that this is an edge case that is likely only triggered by people like us fiddling with the URL bar, I opted to simplify the code and remove it. :grin: 

That said, I'm now realizing that technically it *can* happen if someone expands a section, clicks its permalink, collapses the section, navigates back in their browser history (so the URL no longer contains the hash), and then navigates forward in their browser history. That seems kind of unlikely, but I can re-add the hashchange code if we think it's important. 

> Visually, one pattern that you might be interested in is like GitHub does for wiki pages, where the permalink icon is only visible when hovering/focusing the target.

Yeah, I actually like this pattern in general, but I wasn't sure how the UX would work here because our headers are already used as interactive elements (to expand/collapse their content), so I was worried that complicating the interactivity of that area could make things confusing.

I'm also very wary of hover states because they tend to not work well on touchscreens--I'm actually not sure how/if GitHub wiki's permalink functionality works on tablets, for instance. They also don't usually work very well for keyboard navigation users--for instance, when headings are focused via the tab key, on Firefox there isn't any visual indication that the heading is even focused, let alone that it leads to a permalink. (That said, I might be over-rotating on this accessibility stuff too, lol!)

toolness commented:
Oh! @ottonomy, do the individual strands (explore, build, connect) of the map need their own permalinks too? or just the competencies?

toolness commented:
Ahh, I ran into more funky legit edge cases where hash changes did occur. So I added that hash change code back in!

I also added the funkalicious stack-overflow-esque "make the background a bright color and slowly fade it to the normal color when the user followed a permalink here" behavior. Check it OUT!

http://mozteach.toolness.org/teach-like-mozilla/web-literacy/#credibility

ottonomy commented:
The hashchange feature you added works well. And you're right that it would probably occasionally be triggered by use of the back button. When JS apps can really handle history states, they feel so much more professional. The background color animation works pretty well.

On whether or not the strands need their own permalink too: I could imagine badges being aligned to completion of an entire strand, so if these URLs existed, they might be happily used. I don't know if our project in Oregon would use that feature or not. I think it comes down to a question of the meaning of the lit map. Do you think the strand itself is a node that should be able to be referenced from elsewhere, including badges? Probably. 


I agree with all your reasoning on the visual design and its accessibility of the section icon.

:clap: :clap: :clap: @toolness

toolness commented:
Awesome, thanks @ottonomy!! I think we can make the strands permalinkable, we just have to figure out a good place to put the actual permalink. Perhaps we'll just do the Github Wiki-style permalink-icon-on-hover thing for those?

@mmmavis, while this PR still doesn't have any tests and the design needs to be finalized, I'm interested in your feedback on the actual code--particularly the [`AnchorManagerMixin`](https://github.com/mozilla/teach.webmaker.org/pull/993/files#diff-c8271e258eb867fc3e97253c5cd10bc2) I just added.

toolness commented:
Phew, I think this is ready for review @mmmavis. It ended up being a bit more code than I expected, but I wanted to make sure everything was unit tested. Let me know if you want to walk through this together over vidyo!

toolness commented on mixins/anchor-manager.js:
```diff
@@ -0,0 +1,169 @@
+      attractAttentionToAnchor: true
+    });
+    window.clearTimeout(this.attractTimeout);
+    this.attractTimeout = window.setTimeout(
```
I'm using timeouts here instead of relying on CSS transitions for the same reason we decided to use `TimeoutTransitionGroup` in #745. Basically, I don't trust CSS transition events and I think we should just use timeouts to be safe. :stuck_out_tongue:

mmmavis replied:
Nice! And thanks for bringing up that `TimeoutTransitionGroup` issue again :+1:

toolness commented on pages/web-literacy.jsx:
```diff
@@ -49,7 +49,7 @@ var ActivitiesPage = React.createClass({
           <WebMaps>
             <WebMap head="Explore" subhead="Reading the Web">
-              <Expander head="navigation">
+              <Expander head="navigation" anchorId="navigation">
```
Note that `anchorId` is like a slug-ified version of `head`, so it seems like we might be able to determine it algorithmically to keep things DRY, but [it's a trap](https://www.youtube.com/watch?v=4F4qzPbcFiA) because we will eventually need to localize `head`, while we don't need to localize `anchorId`.

toolness commented on mixins/anchor-manager.js:
```diff
@@ -0,0 +1,169 @@
+var AnchorManagerMixin = {
+  propTypes: {
+    anchorId: React.PropTypes.string,
+    anchorManager: React.PropTypes.object
```
The `anchorManager` prop is only used for testing. I considered making it a [React context](http://www.tildedave.com/2014/11/15/introduction-to-contexts-in-react-js.html) but those are such a pain to test and I kinda want to hold off on using contexts more until their API is stabilized and actually documented in React. We can always migrate to using a context for this later, too.

mmmavis replied:
yea... React context IS confusing :stuck_out_tongue_winking_eye:

mmmavis commented:
This is on my todo list today.  A Vidyo call invite might be sent from me on Monday if I'm unable to finish reviewing this today :wink:

mmmavis commented on components/expander.jsx:
```diff
@@ -1,12 +1,15 @@
     };
   },
-  collapse: function() {
+  collapse: function(e) {
```
is there a reason why we need this change? :open_mouth:

toolness replied:
Oh, no, I think at some point in coding, I had `collapse` double as an event handler, but then i undid that and forgot to remove the args. Will fix.

mmmavis commented on components/expander.jsx:
```diff
@@ -1,12 +1,15 @@
 
+var AnchorManagerMixin = require('../mixins/anchor-manager');
+
 var Expander = React.createClass({
```
@toolness could you add `propTypes` here? :sunglasses:

toolness replied:
Done! Note that `anchorId` isn't in it, because React automatically merges the `propTypes` of mixins and complains if there are any conflicts.

mmmavis replied:
Ah, there's also `children` @toolness :wink:

toolness replied:
Hmm, so I have concerns with how helpful that actually is, since `children` is already a documented React property and the component won't actually crash or anything if `children` isn't defined...  We've actually had react warnings pop up in other components because `children` was declared to be `propTypes.array` instead of `propTypes.node`.

Aside from that, though, because we're using some stub context helper code to create the expander component in our tests, we can't easily add children to the expanders in our tests, and I don't want warnings logged to in our tests complaining that our expanders don't have children...

mmmavis replied:
alright, gotcha!

mmmavis commented on mixins/anchor-manager.js:
```diff
@@ -0,0 +1,169 @@
+AnchorManager.prototype = {
+  register: function(component) {
+    var anchorId = component.props.anchorId;
+    if (process.env.NODE_ENV !== 'production' &&
```
question - why do we have to check `process.env.NODE_ENV`?

toolness replied:
Oh, because it's just a "sanity check" intended for development, and we don't want it to negatively impact production code. React's own [prop validation](http://facebook.github.io/react/docs/reusable-components.html#prop-validation) has similar behavior:

> Note that for performance reasons `propTypes` is only checked in development mode.

mmmavis commented on mixins/anchor-manager.js:
```diff
@@ -0,0 +1,169 @@
+// which is the URL fragment (without the leading hash) of the anchor.
+// This is expected to be constant throughout the lifetime of the
+// component. (If `anchorId` is undefined, the mixin will be largely
+// disabled.)
```
Can we make `anchorId` prop a required prop?  Or is there a reason we are making it optional?

toolness replied:
Yeah, it's because the Clubs Toolkit--which uses the Expander component, which uses this AnchorManagerMixin--isn't using it... while we could (and probably should) add permalinks to the Clubs Toolkit sections, it didn't seem like something that the Expander component should absoutely *require* in order to use, so I made it optional.

mmmavis commented on test/browser/anchor-manager.test.jsx:
```diff
@@ -0,0 +1,261 @@
+
+  var unmountAnchor = function(anchor) {
+    React.unmountComponentAtNode(anchor.getDOMNode().parentNode);
+    anchors.splice(anchors.indexOf(anchor, 1));
```
ummm... why do we need to specifically search from `1`? :no_mouth:

toolness replied:
Oh shoot, this might be a typo--I just wanted to remove the anchor from the list... I think I meant to write this:

```js
anchors.splice(anchors.indexOf(anchor), 1);
```

eep

mmmavis commented on test/browser/anchor-manager.test.jsx:
```diff
@@ -0,0 +1,261 @@
+  });
+
+  afterEach(function() {
+    anchors.slice().forEach(unmountAnchor);
```
I'm curious why we need to create a copy of `anchors` here.  Would `anchors.forEach(unmountAnchor);` be sufficient?

toolness replied:
Oh, so that's because we're iterating over an array while deleting items from it. Technically it *might* work depending on how `forEach` is implemented and/or how we're iterating over it, but I wanted to make sure that we e.g. don't skip any items, so I added the `slice` call.

mmmavis commented on test/browser/anchor-manager.test.jsx:
```diff
@@ -0,0 +1,261 @@
+  it('waits ATTRACT_ATTENTION_TO_ANCHOR_DURATION if defined', function() {
+    var a = renderAnchor({anchorId: 'foo'});
+    a.ATTRACT_ATTENTION_TO_ANCHOR_DURATION = 10;
+    a.attractAttentionToAnchor();
```
What do you think about adding a `a.state.attractAttentionToAnchor.should.be.true;` check here? @toolness

toolness replied:
Sure, just added it!

mmmavis commented:
donnnnne reviewing!  awesome work @toolness :dancer: :dancer: :dancer:
